server.go

```diff
@@ -115,6 +121,71 @@ func ReadOnly() ServerOption {
 	}
 }
 
+// RequestType is the type of Request a client made.
+type RequestType uint
+
+const (
+	Open RequestType = iota
+	Close
+	Read
+	Write
+	Lstat
+	Fstat
+	Setstat
+	Fsetstat
+	Opendir
+	Readdir
+	Remove
+	Mkdir
+	Rmdir
+	Realpath
+	Stat
+	Rename
+	Readlink
+	Symlink
+)
+
+// RequestPacket is information about a client request.
+type RequestPacket struct {
+	Type RequestType
+	// Path is the path the request specified, or if the request specified
+	// a handle instead, Path is the path corresponding to that handle.
+	Path string
+	// TargetPath is the new path in a rename request, or the new path in
+	// a symlink request.
+	TargetPath string
+	// Flags is any flags the request passed.
+	Flags      uint32
+	Attributes *Attributes
+	// Err is the error that occured from handling the request.
+	Err error
+}
+
+// Attributes is optional metadata that may or may not be present in a
+// client request.
+type Attributes struct {
+	Size             *uint64
+	UID              *uint32
+	GID              *uint32
+	Permissions      *fs.FileMode
+	AccessTime       *time.Time
+	ModificationTime *time.Time
+}
+
+// RequestCallback is the type of function called by a Server when
+// WithRequestCallback is set. reqPacket will contain details of
+// a client request the server has recieved.
+type RequestCallback func(reqPacket RequestPacket)
```

Comment:
It makes no sense to pass this struct by value. In fact, since it contains pointers, this mixed-use case of value and reference can easily lead to misunderstanding and false assumptions about the safety of modifying the argument.

If a struct contains a pointer (or a map, or a slice, or an interface, because these also contain pointers) then it should always be passed also by pointer.